Documents/UpgradeExtensibility.md

```diff
@@ -0,0 +1,157 @@
+VFSForGit Upgrade Extensibility Design Document
+===============================
+
+Scope
+-----
+
+Enable groups to extend the Upgrade verb to install / update extra
+components besides just VFSForGit and Git.
+
+Motivation
+----------
+
+VFSForGit can exist as a component of a larger engineering
+system. Some environments could include components that are coupled
+with specific versions of VFSForGit and should be updated with with
+VFSForGit. There could be other interactions specific to the
+environment that need to be accounted for when updating VFSForGit. As
+these requirements are specific to the individual environments
+VFSForGit is being deployed into, we need a way for groups to include
+extra components and logic in the VFSForGit upgrade.
+
+Requirements
+------------
+
+* Groups can include extra components / logic in VFSForGit Upgrade
+
+* Feed needs to support advertising current available version
+
+Non-Goals
+---------
+
+* Initial configuration of the alternate feed URL
+
+* Controlling the entire package of components
+
+Design
+------
+
+VFSForGit will support an alternative "Feed" from which updates can be
+retrieved. When an alternative feed is configured, VFSForGit will
+query this feed for updates instead of the normal feed. This
+alternative feed can contain customized components in addition to the
+VFSForGit and Git installers.
+
+### Feed
+
+The alternative feed will be a NuGet feed. The feed URL is specified
+in the gvfs configuration file (gvfs.config) with the following entry:
+
+    upgrade.feedurl
+
+This is similar to how upgrade rings are specified. The upgrade.feed
+will take precedence over the upgrade ring configuration. When this
+value is configured, VFSForGit will query and download updates from
+this feed. It must be a NuGet feed.
+
+### Upgrade Package Contents
+
+If using the alternative feed, the upgrade package must include
+installers for Git and VFSForGit. It can optionally include installers
+to run before / after installing Git and GVFS. The upgrade logic will
+look in the following locations for the pre / post setup installers:
+
+    PreSetup\install.exe :: Run before the Git and VFSForGit

```

Comment:
Do these have to be `install.exe`? I can imagine engineering system teams preferring something lighter weight like a .cmd/.sh script that could run some installers (to gather things like telemetry), configure some global git config values, set up mergetool, etc

Reply:
If we do stick with `.exe`, we'll need a different extension for MacOS.

Reply:
A script would make some operations easier - I have been thinking about what the right balance is here.

You are right, if we do stick with running an executable, it will have to be tweaked for the different platforms. I think we adjust the extension of the installer that we expect based on the platform being run on, and this would have to follow the same pattern. I will update the document to reflect this.